From 44d62c947a9f0b274f38be6f64386a607b7fbfe1 Mon Sep 17 00:00:00 2001
From: Henrique Silva <jhenrique09.mcz@hotmail.com>
Date: Tue, 31 Jul 2018 14:48:30 -0300
Subject: [PATCH] camera: Allow disabling shutter sound for specific packages

Some third-party cameras play back their shutter sound.
This produces a double sound. This patch fixes this problem.

camera.shutter_sound.blacklist=com.android.camera - allows
to disable the system shutter sound for a specific camera package.

Change-Id: I0a0dd936c9ebf3d2a56aafb7b93a81f1c626a8dc
Signed-off-by: Henrique Silva <jhenrique09.mcz@hotmail.com>
---
 services/camera/libcameraservice/CameraService.cpp | 17 +++++++++++++++++
 services/camera/libcameraservice/CameraService.h   |  1 +
 2 files changed, 18 insertions(+)

diff --git a/services/camera/libcameraservice/CameraService.cpp b/services/camera/libcameraservice/CameraService.cpp
index de0fac5..250e05b 100644
--- a/services/camera/libcameraservice/CameraService.cpp
+++ b/services/camera/libcameraservice/CameraService.cpp
@@ -2096,10 +2096,27 @@ CameraService::Client::Client(const sp<CameraService>& cameraService,
     mRemoteCallback = cameraClient;
 
     cameraService->loadSound();
+    cameraService->ensureCameraShutterSoundDisabled(clientPackageName);
 
     LOG1("Client::Client X (pid %d, id %d)", callingPid, mCameraId);
 }
 
+void CameraService::ensureCameraShutterSoundDisabled(const String16& clientPackageName) {
+    char value[PROPERTY_VALUE_MAX];
+    if (property_get("camera.shutter_sound.blacklist", value, NULL) > 0){
+        std::stringstream disable_shutter_package_list(value);
+        std::string package;
+        while(std::getline(disable_shutter_package_list, package, ',')){
+            if (package.compare(String8(clientPackageName)) == 0){
+                mSoundPlayer[SOUND_SHUTTER] = NULL;
+                mSoundPlayer[SOUND_RECORDING_START] = NULL;
+                mSoundPlayer[SOUND_RECORDING_STOP] = NULL;
+                return;
+            }
+        }
+    }
+}
+
 // tear down the client
 CameraService::Client::~Client() {
     ALOGV("~Client");
diff --git a/services/camera/libcameraservice/CameraService.h b/services/camera/libcameraservice/CameraService.h
index 8d4bcdb..518d71f 100644
--- a/services/camera/libcameraservice/CameraService.h
+++ b/services/camera/libcameraservice/CameraService.h
@@ -178,6 +178,7 @@ public:
     };
 
     void                loadSound();
+    void                ensureCameraShutterSoundDisabled(const String16& clientPackageName);
     void                playSound(sound_kind kind);
     void                releaseSound();
 
-- 
2.7.4

